### PR TITLE
Application stack design

### LINKED ISSUE #191: Watcher Design
We need to have a discussion about the Watcher at a high level since there's a lot of confusion about the vision of what purpose it serves, which leads us to confusion about things like #156.

Rather than tackling individual features ad-hoc, let's talk about how we envision the future of the Watcher.

People have characterized it as:
- A verifier (blocks, withholding, transaction consistency, root chain submissions)
- A full node (non-mining)
- A light node
- An convenience API gateway to the child chain
- A client/wallet

This specifics of what it does and what clients it serves (servers, browsers, wallets, etc.) can lead feature requirements (like whether we want CORS support).

As to actual functional requirements, here's what we need on the integrations end:

### Block explorer APIs (each endpoint should have filtering capabilities):
- Getting all blocks
- Getting a block and its transactions
- Getting all transactions (paginated, per address, per list of addresses)
- Get specific transaction (by id)
- Getting UTXOs (paginated, per address, per list of addresses?)
- Get balance by address

### eWallet APIs:
- Build transaction
- Submit signed transaction
- Get specific transaction (by correlation field)

### Events:
- Subscribe to Websocket event endpoints notifying the following events (probably for a list of addresses?):
  - should exit due to [some fault]
  - Standard Exit (MVP)
    - Exit Challenge
  - In-flight exit (MoreVP)
    - Delegated?
    - Exit challenge
    - Challenge response
  - Exited successfully
  - new transaction submitted
  - new deposit
  - transaction shows up in a block
  - new block
  - ...we probably need to add to this list... 

### UTXO Management
  - Merging / splitting / recommendations

### Security
  - Key management
  - Verify consistency

### Deployment
  - Support multiple instances of watchers per node (many eWallets, many Watchers, 1 plasma node)

This is the beginning of a design discussion. Please drop your opinions in here.

/cc @whoisjeremylam @robinclart @omisego/blockchain @T-Dnzt @Pongch @kevsul @kfichter

### PR DESCRIPTION
**Pull request by kasima**
Draft design document for OmiseGO application stack, informed by #191 and our design call.

Pieces include:
- Watcher
- Informational API Service
- Integration Libraries
- Application Layer

@pnowosie - Could use some help documenting what the watcher currently supports by way of events and APIs.

Any and all additions/comments welcome. 

@pdobacz @whoisjeremylam @T-Dnzt @kevsul @jarindr 

Fixes #191

### CODE REVIEW

File: docs/application_architecture.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+#### Requirements
+
+- Events
+  - Should exit due to fault
+    - Double spend
+    - Invalid exit started
+    - Transaction spent
+    - Out of nowhere transaction created
+    - Block withholding?
+    - ...
```
Comment (kasima) [inline docs/application_architecture.md]: @pnowosie – What does the watcher currently detect and what is planned so far?

File: docs/application_architecture.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  - Successful exit
+  - Transaction in block
+  - New deposit
+  - New transaction
+  - New block
+- API
+  - Submit transaction
+  - Start exit
+  - Challenge exit
+  - ...
```
Comment (kasima) [inline docs/application_architecture.md]: @pnowosie – Same question as above

File: docs/application_architecture.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+#### Design principle
+
+- Provide convenience APIs to proxy to the child chain/root chain/watcher to ease integration and reduce duplicate code in libraries
+- Storage of informational data about the chain
+- Support direct client requests (web browser, mobile, etc.)
+
+#### Requirements
+
+- Events
+  - ...
```
Comment (kasima) [inline docs/application_architecture.md]: @T-Dnzt @jarindr – What events do you need for a first release of the ewallet that aren't covered in the Events listed under *Watcher* ☝️

File: docs/application_architecture.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  - Getting all blocks
+  - Getting a block and its transactions
+  - Getting all transactions (paginated, per address, per list of addresses)
+  - Get specific transaction (by id)
+  - Getting UTXOs (paginated, per address, per list of addresses?)
+  - Get balance by address
+  - Build transaction
+  - Submit signed transaction
+  - Get specific transaction (by correlation field)
+- UTXO management?
```
Comment (kasima) [inline docs/application_architecture.md]: @kevsul @whoisjeremylam – This is up for debate and requires some serious design. With only 4x4 I/O per transaction, without any protocol support, UTXO management seems like it gets very expensive.

File: docs/application_architecture.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
+## OmiseGO Application Architecture
```
Comment (pdobacz) [inline docs/application_architecture.md]: call this `OmiseGO Stack Architecture` perhaps? (if so then rename file too). I'd also add a header of 1-2 sentences about what this doc is and what it's not.

On a related topic: this document omits both the child chain and Ethereum, however I think they sort of belong to the story, as being services that this whole stack talks to. Note that such relation can be established in the document without differentiating into what the `child chain` actually is made of - PoA/PoS etc.

Comment (kasima) [reply docs/application_architecture.md]: Will rename and add child chain and Ethereum to the document. Thanks.

Comment (whoisjeremylam) [reply docs/application_architecture.md]: Under child chain, we should include:
- Fee management: setting of fees, withdrawing fees, possibly burning fees
- Gas management: settings related gas costs for block submissions to Ethereum, process to fund ETH to the operator,

@pdobacz there's more to add here. If you can remember, please add.

Comment (pdobacz) [reply docs/application_architecture.md]: wait, so my understanding is that this "Stack" doesn't include the child chain, it just "talks" to it. 

So it only should be referenced, but not described, to keep this doc pure and not duplicate with, say, `tesuji_design` or other docs on the child chain server.

The two points mentioned belong outside of this client stack, and will be (already are) described in child chain docs.

Comment (whoisjeremylam) [reply docs/application_architecture.md]: Ah I get the point. If this is an 'application layer' stack then a reference to the lower layers is appropriate.

File: docs/application_architecture.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  - In-flight exit started
+  - In-flight exit challenged
+  - Successful exit
+  - Transaction in block
+  - New deposit
+  - New transaction
+  - New block
+- API
+  - Submit transaction
+  - Start exit
```
Comment (kevsul) [inline docs/application_architecture.md]: `Start exit` and `Challenge Exit` are called on the RootChain contract, not the Watcher.
However the Watcher does have a `Get exit data` API call that provides the necessary data for those Rootchain calls

File: docs/application_architecture.md (hunk 5 above)
Comment (kasima) [reply docs/application_architecture.md]: I'm calling the whole thing "Client services." Open to other suggestions.

File: docs/application_architecture.md (hunk 1 above)
Comment (pnowosie) [reply docs/application_architecture.md]: Currently Watcher publishes events in 2 topics
**`transfer:{address}`**
 * `address_received`
 * `address_spent`

**`byzantine`**
 * `block_withholding`
 * `invalid_block`
 * `invalid_exit`
 * `unchallenged_exit`

More details about events and planned ones (WIP) - I've just pushed [branch with `api specs`](https://github.com/omisego/elixir-omg/tree/feature/api-specs-proposal) which @kevsul is taking over

File: docs/stack_architecture.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+- Only include functionality that is critical to the operation of the Plasma security model
+- Strict focus on security role reduces complexity and attack surface area
+- Limited feature helps scalability
+  - The more the watcher does, the slower it can verify
+- 3 primary security functions:
+  - Tracking of the root chain submissions, pulling block contents (from somewhere) and validating, in order to ensure safety of funds passively in possession on the child chain. Watcher notifies in case of the funds are jeopardized.
+  - Proxy API to the child chain API (whatever it may be - PoA server or a P2P PoS network) and the root chain, that ensures that these two are never talked to if the chain is invalid or in unknown state. Only proxy calls that require the chain is operational.
+  - Storage of data critical to access of the funds - UTXO positions, `txbytes` or any other kinds of proofs
+
+#### Requirements
```
Comment (pnowosie) [inline docs/stack_architecture.md]: I think we can refer for details to the (already reviewed) `Specs proposal` via GH pages link.
[Watcher Security Critical spec] (https://omisego.github.io/elixir-omg/#watcher)

File: docs/stack_architecture.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+Non-critical convenience API proxy and provide data about the chain.
+
+#### Design principles
+
+- Provide convenience APIs to proxy to the child chain/root chain/watcher to ease integration and reduce duplicate code in libraries
+- Storage of informational data about the chain
+- Support direct client requests (web browser, mobile, etc.)
+
+#### Requirements
```
Comment (pnowosie) [inline docs/stack_architecture.md]: As above refer to the specs
[Watcher's Informational API](https://omisego.github.io/elixir-omg/#watcher-informational-api-service)

File: docs/stack_architecture.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+#### Design principles
+
+- Adopt all native conventions and standards
+- Encourage open source community development
+
+#### Requirements
+
+- Support all events and API calls of the Watcher
+- Support all events and API calls of the API Service
+
```
Comment (pnowosie) [inline docs/stack_architecture.md]: Shall me mention `omg-js` repo here :question:

**Comment by kevsul** [comment]
Agree with @pnowosie - let's take out the specifics from here and point to the API docs instead.

**Comment by kasima** [comment]
Finally updated to address comments. 🙈 Primarily linked to existing API docs.

**Pull request merged**